common/datavalues/src/columns/primitive/mod.rs

```diff
@@ -189,17 +189,35 @@ impl<T: PrimitiveType> Column for PrimitiveColumn<T> {
         if length == self.len() {
             return Arc::new(self.clone());
         }
-        let iter = self
-            .values()
+
+        let mut res = Vec::<T>::with_capacity(length);
+        let mut offset = 0;
+        let values = self.values();
+
+        const MASK_BITS: usize = 64;
+        for mut mask in filter.values().chunks::<u64>() {
+            if mask == u64::MAX {
+                res.extend(&values[offset..offset + MASK_BITS]);
+            } else {
+                while mask != 0 {
+                    let n = std::intrinsics::cttz(mask) as usize;
+                    res.push(values[offset + n]);
+                    mask = mask & (mask - 1);
+                }
+            }
+            offset += MASK_BITS;
+        }
+
+        for (&v, _) in values
             .iter()
             .zip(filter.values().iter())
+            .skip(offset)
             .filter(|(_, f)| *f)
-            .map(|(v, _)| *v);
+        {
+            res.push(v);
```

Comment:
Processing the remaining in chunks is here, maybe not elegant :(

Reply:
It's not elegant. `skip` may cause extra iterating.